[Immediate] Move to new diagnostic handler model when trying to link modules

xedin opened the pull request:
These changes are trying to fix problem related to changes from upstream
LLVM which changes the way how diagnostics are handled in presence of linker
failures.

Resolves: rdar://problem/35458963

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@swift-ci please clean smoke test

xedin commented:
/cc @jrose-apple my attempt to update Swift to changes in LLVM to resolve CI problems.

slavapestov commented:
@swift-ci please clean smoke test

jrose-apple commented on lib/Immediate/Immediate.cpp:
```diff
@@ -197,9 +171,27 @@ bool swift::immediate::linkLLVMModules(llvm::Module *Module,
+    }
+  };
+
+  Ctx.setDiagnosticHandler(std::unique_ptr<LinkerDiagnosticHandler>(), nullptr);
```
Did you mean `llvm::makeUnique` here? This will create a null unique_ptr.

xedin replied:
Yes, sorry, I had to run and didn't notice that...

jrose-apple reviewed (commented):
This certainly looks like it matches what the old code is doing, but I don't know *why* the old code was doing it. @lhames, any insights?

xedin commented:
@swift-ci please clean smoke test

xedin commented:
@swift-ci please clean smoke test

jrose-apple commented:
Certainly seems good enough to get us building again.

xedin commented:
@jrose-apple Looks like that didn’t help, 11 test failures, can you please take a look? I am AFK right now...

xedin commented:
Ok I'm back and I think I've figured what is going on

xedin committed:
[Immediate] Move to new diagnostic handler model when trying to link modules

These changes are trying to fix problem related to changes from upstream
LLVM which changes the way how diagnostics are handled in presence of linker
failures.

Resolves: rdar://problem/35458963

xedin commented:
@swift-ci please smoke test and merge

xedin commented:
Method simply got renamed to `setDiagnosticHandlerCallBack`.

xedin commented:
@swift-ci please smoke test and merge

rudkx commented:
👏🏻

nkcsgexi commented:
👍

gparker42 commented:
We need a similar fix on the 4.1 branch.
